### DISCUSSION TITLE
How to calculate the group delay

### DISCUSSION THREAD

**Discussion by niubiworker** [NanoComp/meep#2807] [Q&A]
Hello, everyone. How to calculate the group delay or group velocity of a resonant structure (such as a ring resonator)? In some other software, such as Lumerical, I can find a way to calculate group delay, but in Meep, it seems that there is no way to directly calculate group delay. I tried to calculate through definition, but I couldn't find a time monitor in Meep?
 May I ask who can help me calculate the group delay of a resonant structure in meep?

**Comment by stevengj** [comment]
"Group velocity of a resonant structure" is not a thing — group velocity is only a property of propagating modes in translation-invariant/periodic systems like waveguides.  (In which case it is best obtained by a mode solver like MPB.)

I assume you mean that you have a system that is transmitting through a resonant structure, i.e. you have (input port 1)—(resonator)—(output port 2), and you want to compute the group delay from 1–2.

You could use the eigenmode coefficient feature to compute the S12 scattering-matrix element (just compute the output eigenmode coefficient at 2), and do a normalization calculation (as in the waveguide-bend tutorial) to find the input eigenmode coefficient, and divide them to find S12.  (Do not square them!  You want the raw complex amplitudes.)  Then from the phase of S12 at different frequencies you can compute the phase delay dφ/dω.

You can also send a pulse through and monitor the time-domain output.  You can define a [step function](https://meep.readthedocs.io/en/latest/Python_User_Interface/#run-and-step-functions) that does whatever you want at each time step, including outputting the value of the field at some point or saving it to an array.  There is an example of this in the [90° bend tutorial](https://meep.readthedocs.io/en/latest/Python_Tutorials/Basics/#a-90-bend) defining a `get_slice` function that saves a sequence of field slices into an array from every time step.

(Beware that the definition of "group delay" is fundamentally problematic in a lossy system — i.e. if your transmission |S12| is not close to 100%, then it is tricky to define time delay.  Lots of people have been tripped up by this and ended up claiming "superluminal" transmission faster than the speed of light.)

**Reply by niubiworker** [reply]
1. I attempted the operations you suggested, but I think I haven't made my question clear. I currently have a periodic structure (such as ridge waveguide - photonic crystal waveguide - ridge waveguide), and I've solved the band structure through MPB, thus obtaining the modal group velocities. However, I want to know the group velocities of modes during actual transmission. In the documentation you provided, I found a statement saying:["excite the mode using a narrowband pulse and compute the ratio of the Poynting flux to electric-field energy density."](https://meep.readthedocs.io/en/latest/FAQ/#how-do-i-compute-the-group-velocity-of-a-mode) My understanding is to place a monitor (point monitor?) on the output ridge waveguide and then perform the calculation (with possible code as follows), but this gives only a single value. I now wish to determine the relationship between group velocity and wavelength changes. How should I proceed? (Should I take an average value if it's an area monitor?)
2. I have also tried using modal decomposition, but the group index I obtained is the same as that of the fundamental mode (greater than 2.3, consistent with the refractive index of the material). How can I achieve the group index of my slow-light waveguide through modal decomposition?
3. I currently have an unused 3060; can I use the GPU to accelerate Meep calculations?
Thank you very much.

```
...
structure=xxxxx
sim=xxxx
sim.run()
poynting_Vector_{x,y,z}=sim.get_array(component=mp.S{x,y,z},center=xxx,size=xxx)
Energydensity=sim.get_array(component=mp.EnergyDensity,center=xxx,size=xxx)
.....
```

**Comment by stevengj** [comment]
(I feel like @oskooi wrote a tutorial somewhere that includes a computation of the transmitted or reflected phase φ(ω), but I can't find it right now.)

**Reply by oskooi** [reply]
I think @stevengj is referring to [Tutorial/Phase of a Total Internal Reflected Mode](https://meep.readthedocs.io/en/latest/Python_Tutorials/Mode_Decomposition/#phase-of-a-total-internal-reflected-mode) which demonstrates how to compute the complex reflection coefficient.

**Comment by stevengj** [comment]
> I now wish to determine the relationship between group velocity and wavelength changes.

Why not use the MPB calculation for that?  Doing it in time domain is way less efficient.

> I've solved the band structure through MPB, thus obtaining the modal group velocities. However, I want to know the group velocities of modes during actual transmission.

Why do you think you'll get a different answer "during actual transmission" as opposed to from a modal calculation?

>  My understanding is to place a monitor (point monitor?) on the output ridge waveguide and then perform the calculation (with possible code as follows), but this gives only a single value.

No, you would do a broadband calculation with a pulse, and then use a DFT monitor to get the Fourier-transformed fields at each frequency of interest.   But this should be redundant with the MPB modal calculation, assuming you are exciting a single mode at each frequency.  (And if you aren't exciting a single mode at each frequency, your group-velocity calculation will be broken.)
